### PR TITLE
Document new json operators @>, <@, ?, ?|, ?&, #> and #>>

### LINKED ISSUE #1433: Document: feat(expr): support jsonb `@>`, `<@`, `?`, `?|` and `?&` operator
This issue tracks the documentation update needed for the merged PR #13056.

Source PR URL: https://github.com/risingwavelabs/risingwave/pull/13056
Source PR Merged At: 2023-10-30T04:37:04Z

### LINKED ISSUE #1436: Document: feat(expr): support `#>` and `#>>` operator for extracting jsonb at a path
This issue tracks the documentation update needed for the merged PR #13110.

Source PR URL: https://github.com/risingwavelabs/risingwave/pull/13110
Source PR Merged At: 2023-10-30T05:56:32Z

### PR DESCRIPTION
**Pull request by neverchanje**
## Info

- **Description**

  - [ What's changed? Which parts of the docs are affected? ]

- **Notes**

  Examples are copied from sqllogictests.

- **Related code PR**

  - https://github.com/risingwavelabs/risingwave/pull/13056
  - https://github.com/risingwavelabs/risingwave/pull/13110

- **Related doc issue**
  
  Resolves https://github.com/risingwavelabs/risingwave-docs/issues/1433
  Resolves https://github.com/risingwavelabs/risingwave-docs/issues/1436

## For reviewers

- **Preview**

  - [ Paste the preview link to the updated page(s) here. Edit this item after the preview site is ready. To find the updated pages, scroll down to locate and open the Amplify preview link and select the **dev** version of the documentation. ]

- **Key points**

  - [ Parts that may need revision or extra consideration. ]

## Before merging

- [ ] I have checked the doc site preview, and the updated parts look good.

- [ ] I have acquired the approval from the owner (and optionally the reviewers) of the code PR and at least one tech writer (`CharlieSYH`, `emile-00`, & `hengm3467`).

### CODE REVIEW

File: docs/sql/functions-operators/sql-function-json.md [hunk 1]
```diff
@@ -228,6 +228,118 @@ SELECT '{"a": "b"}'::jsonb || '42'::jsonb;
...
+'{"product": "PostgreSQL", "version": 9.4, "jsonb": true}'::jsonb @> '{"version": 9.4}'::jsonb → t
+
+'{"foo": {"bar": "baz"}}'::jsonb @> '{"bar": "baz"}'::jsonb → f
+
+'{"foo": {"bar": "baz"}}'::jsonb @> '{"foo": {}}'::jsonb → t
+```
+
+### `jsonb <@ jsonb -> boolean`
+
+This operator checks if the left `jsonb` value is contained at the top level within the right `jsonb` value.
```
Comment (wangrunji0408) [inline docs/sql/functions-operators/sql-function-json.md]: I suggest to mention that this is a reversed version of `@>` so they share the same behavior in detail.

File: docs/sql/functions-operators/sql-function-json.md [hunk 2]
```diff
@@ -228,6 +228,118 @@ SELECT '{"a": "b"}'::jsonb || '42'::jsonb;
...
+
+```sql title=Examples
+'[1, 2, 3]'::jsonb @> '[1, 3]'::jsonb → t
+
+'{"product": "PostgreSQL", "version": 9.4, "jsonb": true}'::jsonb @> '{"version": 9.4}'::jsonb → t
+
+'{"foo": {"bar": "baz"}}'::jsonb @> '{"bar": "baz"}'::jsonb → f
+
+'{"foo": {"bar": "baz"}}'::jsonb @> '{"foo": {}}'::jsonb → t
+```
```
Comment (wangrunji0408) [inline docs/sql/functions-operators/sql-function-json.md]: This function has a lot of details in corner cases.
https://github.com/risingwavelabs/risingwave/blob/403dcc9d2461d3b8823e282ba14f742beda2b73e/src/expr/impl/src/scalar/jsonb_contains.rs#L23
Do we need to mention them one by one?

**Review by wangrunji0408 — COMMENTED** [review]
Generally LGTM!

File: docs/sql/functions-operators/sql-function-json.md (hunk 2 above)
Comment (neverchanje) [reply docs/sql/functions-operators/sql-function-json.md]: No. 😢 We just document primary usages.

File: docs/sql/functions-operators/sql-function-json.md (hunk 1 above)
Comment (neverchanje) [reply docs/sql/functions-operators/sql-function-json.md]: The suggestion is good but I think it's straightforward to find that <@ is a reversed version of @>. I prefer keeping the original way. @hengm3467 Please help refine the doc and see if we need to use better phrasing.

File: docs/sql/functions-operators/sql-function-json.md [hunk 3]
```diff
@@ -228,6 +228,118 @@ SELECT '{"a": "b"}'::jsonb || '42'::jsonb;
 [{"a": "b"}, 42]
 ```
 
+### `jsonb @> jsonb -> boolean`
+
+This operator checks if the left `jsonb` value contains the right `jsonb` value at the top level.
```
Comment (xiangjinwu) [inline docs/sql/functions-operators/sql-function-json.md]: Where does the phrase 'at the top level' come from?

In [PostgreSQL](https://www.postgresql.org/docs/16/datatype-json.html#JSON-CONTAINMENT) wordings, containment (`@>` `<@`) is **nested** while existence (`?` `?|` `?&`) is **at the top level**.

**Review by xiangjinwu — APPROVED** [review]
rest lgtm

File: docs/sql/functions-operators/sql-function-json.md (hunk 3 above)
Comment (neverchanje) [reply docs/sql/functions-operators/sql-function-json.md]: TBH, it's first generated by gpt4 and modified by me according to the postgres official doc: https://www.postgresql.org/docs/9.5/functions-json.html

@> | Does the left JSON value contain the right JSON path/value entries **at the top level?**

Here is how it states. GPT-4 didn't add "top level". I did. I am not sure which one is true now...

Comment (neverchanje) [reply docs/sql/functions-operators/sql-function-json.md]: @xiangjinwu I wonder if you can help test the behavior and correct the document.

Comment (xiangjinwu) [reply docs/sql/functions-operators/sql-function-json.md]: I see, so basically PostgreSQL 16 rephrased its doc from 9.5.

> which one is true now

Both are true. Existence operators only check at top level, while containment operators check all nested levels, from the top to the bottom.

PostgreSQL doc spends a lot of examples to define what it means for a json (tree) to **contain** another json (tree). If we want to be concise and not replicating PostgreSQL, can we link to it?

File: docs/sql/functions-operators/sql-function-json.md [hunk 4]
```diff
@@ -228,6 +228,118 @@ SELECT '{"a": "b"}'::jsonb || '42'::jsonb;
 [{"a": "b"}, 42]
 ```
 
+### `jsonb @> jsonb -> boolean`
```
Comment (hengm3467) [inline docs/sql/functions-operators/sql-function-json.md:231]: These are all operators. I think we want to have a dedicated section for operators, and list these operators in this new section.

**Review by hengm3467 — APPROVED** [review]

File: docs/sql/functions-operators/sql-function-json.md (hunk 3 above)
Comment (hengm3467) [reply docs/sql/functions-operators/sql-function-json.md]: Good idea. Added a link to this section.

Comment (hengm3467) [reply docs/sql/functions-operators/sql-function-json.md]: And removed "top level" related descriptions from the containment operators.

**Pull request merged**